src/lib/resolver.js

```diff
@@ -34,42 +33,35 @@ export function createEntry(node, lang) {
     if (!attrs) {
       attrs = Object.create(null);
     }
-    attrs[key] = createAttribute(node[key], lang, node.$i + '.' + key);
+    attrs[key] = createAttribute(node[key]);
   }
 
-  let entity = {
+  return {
     value: node.$v !== undefined ? node.$v : null,
     index: node.$x || null,
     attrs: attrs || null,
   };
-
-  meta.set(entity, {id: node.$i, lang});
-  return entity;
 }
 
-function createAttribute(node, lang, id) {
+function createAttribute(node) {
   if (typeof node === 'string') {
     return node;
   }
 
-  let attr = {
+  return {
     value: node.$v || (node !== undefined ? node : null),
     index: node.$x || null,
   };
-
-  meta.set(attr, {id, lang});
-  return attr;
 }
 
 
-export function format(ctx, args, entity) {
+export function format(ctx, lang, args, entity) {
   if (typeof entity === 'string') {
     return [{}, entity];
   }
 
   if (resolutionChain.has(entity)) {
-    const m = meta.get(entity);
-    throw new L10nError('Cyclic reference detected: ' + m.id, m.id, m.lang);
+    throw new L10nError('Cyclic reference detected');
```

Comment:
The only thing we're losing is that this error might happen in a different entity than the currently formatted one, and `err.id` will be the name of the latter.